lib/invoker.js

```diff
@@ -23,7 +23,7 @@ class Invoker {
     Invoker.prepareProcessor(args, asciidoctor)
     const options = this.options.options
     if (this.options.stdin) {
-      Invoker.convertFromStdin(options, args)
+      await Invoker.convertFromStdin(options, args)
```

Comment:
It sounds like nothing else is performed by the function afterwards.

So is the `await` statement required?

Reply:
That's true but the method `invoke` can be called in downstream project. So we should mark the function as `async` right?

```js
class AsciidoctorRevealJsInvoker extends Invoker {

  async invoke() {
    await super.invoke()
    // do something
  }
}
```

Reply:
My understanding is that the `await` is about the "returned result" of a function.

`await super.invoke()` is equivalent as `super.invoke().then(() => ... do the rest).

But I have a doubt now, and I would need to check to be 100% of what I just wrote.

Reply:
> If async and await remain, I think I need to do a major/breaking release?

Actually, the current behavior was undefined because the `convertFromStdin` was using a callback, so:

```js
class AsciidoctorRevealJsInvoker extends Invoker {

  invoke() {
    super.invoke()
    // no guarantee that convertFromStdin has completed... :(
  }
}
```

Now that we are using a Promise we can `await`:

```js
class AsciidoctorRevealJsInvoker extends Invoker {

  async invoke() {
    await super.invoke()
    // convertFromStdin has completed!
  }
}
```

But it does not require a breaking release because the following is still working:

```js
class AsciidoctorRevealJsInvoker extends Invoker {

  // even if invoke is declared as async we don't need to explicitly add the async keyword
  invoke() {
    super.invoke()
  }
}
```

Reply:
I think it's needed if we want to make sure that "convertFromStdin" has completed, see: https://gist.github.com/Mogztter/ae6ee4b62dbed7461f43471d6040870b

Reply:
@oncletom does it make sense? am I missing something?

Reply:
I did try your gist, and here is what I come with: https://gist.github.com/oncletom/f5583337eae10e8a80734d6853304999

_Returning_ a value is what made a difference. So using `async` is only needd if you directly states `await` within the body of the function. You can combine promises and async/await — what matters is "is the resolved result returned?".

Reply:
So if I understand correctly I either need to add `await` or return the Promise. That way downstream can await the promise? 

What is the best approach here? I don't really have a preference...

Reply:
Returning makes more sense to me — mainly because you forward the handling decision to the consuming function, instead of "blocking" while you await.

One thing I realised yesterday, is labelling a function as `async` wraps its result in a promise anyway.


```js
async function test() {
  return 2
}
```

This function returns a promise, that resolves to `2`.
If a function returns a promise, I don't think there is a need to make it `async`, unless you want to use `await` directly inside it.

Reply:
Ok.

> I don't think there is a need to make it async, unless you want to use await directly inside it.

I was using the keyword `async` as an hint/documentation since it's a public API. 
I believe that `await` is not mandatory when you return a Promise but I also believe that the result is the same:

```js
invoke() {
  return Invoker.convertFromStdin()
}

// is equivalent to
async invoke() {
  return await Invoker.convertFromStdin()
}
```

Even if I don't explicitly use `async` and `await` it will be "blocking" until the Promise is resolved (or rejected) since it's the returned value? Am I wrong?

Reply:
Reading the code again, I think that we should exit the process after `convertFromStdin` to be consistent with the other branches. In this case, we should also check the failure level and exit with the appropriated code.

We might want to introduce a new `function` or a callback if the consumer of this API wants to execute some code after the Asciidoctor processing but before the `process.exit(code)`.

Reply:
> I was using the keyword async as an hint/documentation since it's a public API.

Totally legit, it's a good idea as it makes it explicit 👍

Reply:
Alright, let's do it!